File: src/replaceSpace.cpp
class Solution {
public:
	void replaceSpace(char *str,int length) {
        // 复杂度 O(N)

        if (str == NULL || length < 0) {
            return;
        }

        // 统计空格的个数，记录下总长度
        int spaces = 0;
        int i = 0;
        int reallen = 0;
        while (str[i] != '\0') {
            ++reallen;
            if (str[i] == ' '){
                ++spaces;
            }
            ++i;
        }

        // 新的数组空间大小
        int newlen = reallen + 2 * spaces;

        // 双指针，从后往前写，不需要开辟另外一个数组
        int p1 = reallen;
        int p2 = newlen;
        while (p1 >= 0 && p2 > p1) {
            if (str[p1] == ' ') {
                str[p2--] = '0';
                str[p2--] = '2';
                str[p2--] = '%';
            } else {
                str[p2--] = str[p1];
            }
            --p1;
        }
	}
};

File: src/java/ReplaceSpace.java
public class Solution {
    public String replaceSpace(StringBuffer str) {

        // corner case
        if (str == null) {
            return null;
        }

        // calculate the space number
        int spaceNum = 0;
        for (int i = 0; i < str.length(); i++) {
            if (str.charAt(i) == ' ') {
                spaceNum++;
            }
        }

        // resize the stringbuffer str
        int oldindex = str.length() - 1;
        int newlength = str.length() + 2 * spaceNum;
        int newindex = newlength - 1;
        str.setLength(newlength);

        // copy and replace
        for (; oldindex >= 0; oldindex--) {
            if(str.charAt(oldindex) == ' '){
                str.setCharAt(newindex--, '0');
                str.setCharAt(newindex--, '2');
                str.setCharAt(newindex--, '%');
            }else{
                str.setCharAt(newindex--, str.charAt(oldindex));
            }
        }

        return str.toString();
    }
}
